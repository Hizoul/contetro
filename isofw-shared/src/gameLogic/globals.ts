
export interface IPosition {
  x: number, y: number
}

export type Block = IPosition[]

const tetromino = {
  i: 0, o: 1, t: 2, s: 3, l: 4
}

const rotations = {
  normal: 0, right: 1, upsideDown: 2, left: 3, mirrorNormal: 4, mirrorRight: 5, mirrorUpsideDown: 6, mirrorLeft: 7
}

const logType = {
  rolled: 0, placed: 1, skipped: 2, considering: 3
}

const skipReason = {
  unplacable: 0, blockUnavailable: 1
}

const GAMEFIELD_MAX_WIDTH = 10
const GAMEFIELD_MAX_HEIGHT = 20

const LOCAL_COLORS = ["#1273de", "#b80000", "#008b02", "#fccb00"]

export {
  tetromino, rotations, logType, skipReason, GAMEFIELD_MAX_WIDTH, GAMEFIELD_MAX_HEIGHT,
  LOCAL_COLORS
}
