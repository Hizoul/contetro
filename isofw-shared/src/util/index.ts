const mulberry32 = (a: number) => {
  return () => {
    let t = a += 0x6D2B79F5
    t = Math.imul(t ^ t >>> 15, t | 1)
    t ^= t + Math.imul(t ^ t >>> 7, t | 61)
    return ((t ^ t >>> 14) >>> 0) / 4294967296
  }
}

const random = {
  function: mulberry32(Math.floor((Math.random() * Number.MAX_SAFE_INTEGER) + Number.MIN_SAFE_INTEGER)),
  setSeed: (seed: any) => {
    random.function = mulberry32(seed)
  }
}

const randomInRange = (min: number, max: number) => {
  return Math.floor((random.function() * max) + min)
}

const randomColor = () => {
  return `rgba(${randomInRange(0, 255)},${randomInRange(0, 255)},${randomInRange(0, 255)},1)`
}

const range = (from: number, to: number, step: number = 1) => {
  const rangeArray = []
  for (let i = from; i < to; i += step) {
    rangeArray.push(i)
  }
  return rangeArray
}

export {
  randomInRange, range, random, mulberry32, randomColor
}
