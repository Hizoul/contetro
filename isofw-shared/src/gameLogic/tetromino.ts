import { IPosition } from "./globals"
import tetrominoShapes from "./shapes"

export class Tetromino {
  public position: IPosition
  public type: number
  public rotation: number
  public from: string
  public touchingBlocks: Tetromino[]
  private checkedBlocks: Tetromino[]
  constructor(type: number, rotation: number, position: IPosition) {
    this.type = type
    this.rotation = rotation
    this.position = position
    this.touchingBlocks = []
    this.checkedBlocks = []
    this.from = ""
  }
  public getBlock() {
    return tetrominoShapes[this.type][this.rotation]
  }
  public getTranslatedBlock() {
    const block = this.getBlock()
    if (block != null) {
      const newBlock = []
      for (const item of block) {
        newBlock.push({
          x: this.position.x + item.x,
          y: this.position.y + item.y
        })
      }
      return newBlock
    }
    return block
  }
  public addTouchingBlock(block: Tetromino) {
    if (this.touchingBlocks.indexOf(block) === -1) {
      this.touchingBlocks.push(block)
    }
  }
  public getConnectedBlocks() {
    this.checkedBlocks = []
    for (const block of this.touchingBlocks) {
      this.checkBlock(block)
    }
    return this.checkedBlocks
  }
  private checkBlock(block: Tetromino) {
    if (this.checkedBlocks.indexOf(block) === -1) {
      this.checkedBlocks.push(block)
      for (const childBlock of block.touchingBlocks) {
        this.checkBlock(childBlock)
      }
    }
  }
}
