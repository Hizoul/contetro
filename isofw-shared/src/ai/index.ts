import { Tetromino } from "../gameLogic/tetromino"
import { GameField } from "../gameLogic/gameField"
import { GAMEFIELD_MAX_WIDTH } from "../gameLogic/globals";
import { GameScore } from "../gameLogic/gameScore";

const SCORE_MODIFIER = 0.01;

const heuristic_touching = (field: GameField, player: string) => {
  let score = 0
  const playerBlocks = field.log.getBlocksPerPlayer(field)
  let i = 0
  playerBlocks[player].forEach((block) => {
    playerBlocks[player].forEach((compareBlock) =>
    block.getTranslatedBlock().map((part) => compareBlock.getTranslatedBlock().map((comparePart) => {
      if (part.x === comparePart.x + 1 && part.y === comparePart.y ||
        part.x === comparePart.x - 1 && part.y === comparePart.y ||
        (part.x === comparePart.x) &&
        (part.y === comparePart.y - 1 || part.y === comparePart.y + 1)) {
        return playerBlocks[player][i].addTouchingBlock(compareBlock)
      }
    })))
    i++
  })
  const alreadyCounted: Tetromino[] = []
  playerBlocks[player].forEach((block) => {
    if (alreadyCounted.indexOf(block) === -1 &&
      block.touchingBlocks.length > 0) {
        const connected = block.getConnectedBlocks()
        connected.forEach((conblock) => alreadyCounted.push(conblock))
        score += connected.length
    }
  })
  return score
}


const heuristic_connectability = (field: GameField, player: string) => {
  let score = 0
  const playerBlocks = field.log.getBlocksPerPlayer(field)
  const lowestField = field.getLowestFromField()
  playerBlocks[player].forEach((block) => {
    const block_positions = block.getTranslatedBlock()
    block_positions.forEach((pos) => {
      let current_y = lowestField[pos.x];
      let y_cmp = pos.y + 1
      if (current_y === y_cmp) {
        score += SCORE_MODIFIER
      }
      if (pos.x -1 >= 0) {
        let left_y = lowestField[pos.x -1]
        if (left_y <= pos.y) {
          score += SCORE_MODIFIER
        }
      }
      if (pos.x + 1 < GAMEFIELD_MAX_WIDTH) {
        let right_y = lowestField[pos.x +1]
        if (right_y <= pos.y) {
          score += SCORE_MODIFIER
        }
      }
    })
  })
  return score
}

const heuristic_enemy_block = (field: GameField, player: string) => {
  let score = 0
  const playerBlocks = field.log.getBlocksPerPlayer(field)
  const lowestField = field.getLowestFromField()
  const player_index = field.players.indexOf(player)
  const other_player_index = (player_index === 1) ? 0 : 1
  const other_player = field.players[other_player_index]
  console.log("GOT PLAYER", player, other_player, playerBlocks)
  playerBlocks[player].forEach((block) => {
    const player_block = block.getTranslatedBlock()
    playerBlocks[other_player].forEach((block) => {
      const opponent_block = block.getTranslatedBlock()
      player_block.forEach((p1) => {
        opponent_block.forEach((p2) => {
          if (((p1.y == p2.y) && (p1.x == (p2.x + 1) || 
          p1.x == (p2.x - 1))) ||
          ((p1.x == p2.x) && p1.y == (p2.y + 1))) {
            score += SCORE_MODIFIER
          }
        })
      })
    })
  })
  return score
}


const get_best_heuristic_play = (field: GameField, player: string) => {
  return get_best_heuristic_play_for_params(field, player, 0.3, 0.5, 1.0, 20.0)
}

const get_best_heuristic_play_for_params = (field: GameField, player: string, connectability_weight: number, enemy_block_weight: number, touching_weight: number, score_weight: number) => {
  let current_best_heu = -99999999999;
  let best_play_index = -1;
  field.getPossiblePlays().forEach((possible_play) => {
    const future_field = field.clone()
    const placed = future_field.placeBlockUsingPlay(possible_play.globalIndex)
    const gameScore = new GameScore(future_field)
    if (!placed) {
      console.log("Couldn't make a possible play, somethings wrong!")
    }
    let heuristic_value = 0
    heuristic_value += heuristic_connectability(future_field, player) * connectability_weight
    heuristic_value += heuristic_enemy_block(future_field, player) * enemy_block_weight
    heuristic_value += heuristic_touching(future_field, player) * touching_weight
    heuristic_value += gameScore.getScoreForPlayer(player) * score_weight
    if (heuristic_value > current_best_heu) {
      current_best_heu = heuristic_value
      best_play_index = possible_play.globalIndex
    }
  })
  return best_play_index
}

export {
  heuristic_connectability, heuristic_touching, heuristic_enemy_block, get_best_heuristic_play, get_best_heuristic_play_for_params
}