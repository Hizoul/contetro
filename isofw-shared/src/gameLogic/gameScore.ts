import { GameField } from "./gameField"
import { Tetromino } from "./tetromino"

export class GameScore {
  public gameField: GameField
  constructor(gameField: GameField) {
    this.gameField = gameField
  }

  public getScoreForPlayerAtIndex(index: number) {
    return this.getScoreForPlayer(this.gameField.players[index])
  }

  public getScoreForPlayer(player: string) {
    let score = 0
    const playerBlocks = this.gameField.log.getBlocksPerPlayer(this.gameField)
    let i = 0
    playerBlocks[player].forEach((block) => {
      playerBlocks[player].forEach((compareBlock) =>
      block.getTranslatedBlock().map((part) => compareBlock.getTranslatedBlock().map((comparePart) => {
        if (part.x === comparePart.x + 1 && part.y === comparePart.y ||
          part.x === comparePart.x - 1 && part.y === comparePart.y ||
          (part.x === comparePart.x) &&
          (part.y === comparePart.y - 1 || part.y === comparePart.y + 1)) {
          return playerBlocks[player][i].addTouchingBlock(compareBlock)
        }
      })))
      i++
    })
    const alreadyCounted: Tetromino[] = []
    playerBlocks[player].forEach((block) => {
      if (alreadyCounted.indexOf(block) === -1 &&
        block.touchingBlocks.length > 0) {
          const connected = block.getConnectedBlocks()
          if (connected.length >= 3) {
            connected.forEach((conblock) => alreadyCounted.push(conblock))
            score += connected.length
          }
      }
    })
    const allPositions = this.gameField.gatherAllPositions()
    playerBlocks[player].forEach((block) => {
      let maxSub = 0
      block.getTranslatedBlock().map((part) => {
        let highestY = 0
        let foundPart = false
        allPositions.forEach((comparePart) => {
          if ((part.x === comparePart.x && comparePart.y < part.y) && highestY <= comparePart.y) {
            foundPart = true
            highestY = comparePart.y + 1
          }
        })
        if (highestY > 0 || !foundPart) {
          maxSub += highestY - part.y
        }
      })
      if (maxSub < -2) { // TODO: make configurable
        maxSub = -2
      }
      score += maxSub
    })
    return score
  }
  public getWinner() {
    let players = this.gameField.players
    let maxScore = -99
    let isDraw = false
    let winner = null
    for (let player of players) {
      let playerScore = this.getScoreForPlayer(player)
      if (maxScore < playerScore) {
        isDraw = false
        winner = player
        maxScore = playerScore
      } else if (maxScore == playerScore) {
        isDraw = true
        winner = null
      }
    }
    return {
      maxScore, winner, isDraw
    }
  }
}
