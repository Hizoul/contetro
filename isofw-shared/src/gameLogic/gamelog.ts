import { GameField } from "./gameField"
import { logType } from "./globals"
import { Tetromino } from "./tetromino"

export interface IGameLogItem {
  type: number
  payload: any
}

export class GameLog {
  public log: IGameLogItem[]
  constructor(previousLog?: IGameLogItem[]) {
    this.log = Array.isArray(previousLog) ? previousLog : []
  }
  public addItem(item: IGameLogItem) {
    this.log.push(item)
  }

  public blockPlaced(user: string, blockType: number, blockOrientation: number, x: number, y: number) {
    this.addItem({
      type: logType.placed,
      payload: {
        from: user,
        block: blockType,
        orientation: blockOrientation,
        x, y
      }
    })
  }

  public rolledBlock(user: string, block: number) {
    this.addItem({
      type: logType.rolled,
      payload: {
        from: user,
        block
      }
    })
  }

  public playerSkipped(user: string, block: number, reason: number) {
    this.addItem({
      type: logType.skipped,
      payload: {
        from: user, block, reason
      }
    })
  }

  public createConsideration(play_index: number = 0) {
    this.addItem({
      type: logType.considering,
      payload: {
        play_index
      }
    })
  }

  public getBlocks() {
    const blocks: Tetromino[] = []
    this.log.forEach((item) => {
      if (item.type === logType.placed) {
        const tetromino = new Tetromino(item.payload.block, item.payload.orientation, {
          x: item.payload.x,
          y: item.payload.y
        })
        tetromino.from = item.payload.from
        blocks.push(tetromino)
      }
    })
    return blocks
  }

  public getBlocksPerPlayer(gameField: GameField) {
    const playerBlocks: {[index: string]: Tetromino[]} = {}
    gameField.players.forEach((player) => playerBlocks[player] = [])
    for (const item of this.log) {
      if (item.type === logType.placed) {
        playerBlocks[item.payload.from].push(new Tetromino(item.payload.block, item.payload.orientation, {
          x: item.payload.x,
          y: item.payload.y
        }))
      }
    }
    return playerBlocks
  }

  public getLatestRolled() {
    let i
    i = this.log.length - 1
    while (i >= 0) {
      if (this.log[i].type === logType.rolled) {
        return this.log[i]
      }
      i--
    }
    return null
  }

  public getCurrentBlock() {
    const roll: any = this.getLatestRolled()
    if (roll != null) {
      return roll.payload.block
    }
  }

  public getCurrentPlayer() {
    const roll: any = this.getLatestRolled()
    if (roll != null) {
      return roll.payload.from
    }
  }

  public getCurrentPlay() {
    let entry
    entry = this.log[this.log.length - 1]
    if (entry.type === logType.considering) {
        return entry.payload.play_index
    }
    return 0
  }

  public getBlockCount(player: string) {
    const blocks = [5, 5, 5, 5, 5]
    this.log.forEach((entry) => {
      if (entry.type === logType.placed && entry.payload.from === player) {
        blocks[entry.payload.block]--
      }
    })
    return blocks
  }
}
