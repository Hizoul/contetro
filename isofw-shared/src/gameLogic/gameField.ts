import { jsToRustLog } from "isofw-shared/src/util/rustLogParser"
import { cloneDeep, findIndex } from "lodash"
import { randomInRange, range } from "../util"
import { GameLog } from "./gamelog"
import { GameScore } from "./gameScore"
import {
  GAMEFIELD_MAX_HEIGHT, GAMEFIELD_MAX_WIDTH, IPosition,
  logType, skipReason, tetromino
} from "./globals"
import { amountRotations } from "./shapes"
import { Tetromino } from "./tetromino"

export interface IPlayer {
  name: string
}
export interface IPlay {
  x: number
  rotation: number
  block: number
  globalIndex: number
}

export class GameField {
  public log: GameLog
  public players: string[]
  public playerColors: {[index: string]: string}
  public blockRestrictions: {[index: string]: number[]}
  public gameOver: boolean
  public blocks: Tetromino[]
  public controlType: {[index: string]: number}
  public deterministic_mode: boolean
  private rollIndex: number
  constructor(players: string[], log?: GameLog, deterministic_mode: boolean = false) {
    this.players = players
    this.playerColors = {}
    this.blocks = []
    this.gameOver = false
    this.deterministic_mode = deterministic_mode
    this.blockRestrictions = {}
    if (log === null || log === undefined) {
      this.log = new GameLog()
      this.initBlockRestrictions()
    } else {
      // required so line 40 is satisfied by ts
      this.log = log
      this.restoreFromLog(log)
    }
    if (this.log.getCurrentPlayer() == null) {
      if (this.deterministic_mode) {
        this.log.rolledBlock(this.players[0], 0)
        this.log.createConsideration()
      } else {
        this.randomNextPlayer()
      }
    }
    this.rollIndex = 0
    this.controlType = {}
    for (const p of players) {
      this.controlType[p] = 0
    }
  }

  public initBlockRestrictions() {
    this.players.forEach((player) => this.blockRestrictions[player] = [5, 5, 5, 5, 5])
  }

  public getHighestY(xpos: number, block: Tetromino) {
    let sum: number = 0
    let yVal: number = 0
    block.position.x = xpos
    block.position.y = 0
    const parts = block.getTranslatedBlock()
    const lowestYs: number[] = []
    for (const part of parts) {
      if (lowestYs[part.x] == null) {
        lowestYs[part.x] = 99
      }
      if (lowestYs[part.x] > part.y) {
        lowestYs[part.x] = part.y
      }
    }
    const lowestField = this.getLowestFromField()
    for (const part of parts) {
      if (lowestField[part.x] != null) {
        sum = lowestField[part.x] - lowestYs[part.x]
      } else {
        sum = -lowestYs[part.x]
      }
      if (sum > yVal) {
        yVal = sum
      }
    }
    return yVal
  }

  public gatherAllPositions() {
    const allPositions: IPosition[] = []
    this.blocks.forEach((block) => block.getTranslatedBlock().forEach((pos) => allPositions.push(pos)))
    return allPositions
  }

  public getLowestFromField() {
    const values: number[] = []
    this.blocks.forEach((block) => block.getTranslatedBlock().forEach((pos) => {
      if (values[pos.x] == null) {
        values[pos.x] = 0
      }
      if (values[pos.x] <= pos.y) {
        values[pos.x] = pos.y + 1
      }
    }))
    return values
  }

  public getHoverBlock() {
    let y
    const play = this.getPossiblePlays()[this.log.getCurrentPlay()]
    if (play != null) {
      y = this.getHighestY(play.x, new Tetromino(play.block, play.rotation, {
        x: play.x,
        y: 0
      }))
      const blockInstance = new Tetromino(play.block, play.rotation, {
        x: play.x,
        y
      })
      blockInstance.from = this.log.getCurrentPlayer()
      return blockInstance
    }
    return null
  }

  public canPlaceBlock(block: Tetromino) {
    let current_player = this.log.getCurrentPlayer()
    if (current_player == null || this.blockRestrictions[current_player][block.type] <= 0) {
      return false
    }
    if (block != null) {
      const ypos = this.getHighestY(block.position.x, block)
      block.position.y = ypos
      const parts = block.getTranslatedBlock()
      for (const part of parts) {
        if (part.y >= GAMEFIELD_MAX_HEIGHT || part.y < 0) {
          return false
        }
        if (part.x >= GAMEFIELD_MAX_WIDTH || part.x < 0) {
          return false
        }
      }
      const positions = this.gatherAllPositions()
      for (const position of positions) {
        for (const part of parts) {
          if (part.x === position.x && part.y === position.y) {
            return false
          }
        }
      }
      return true
    }
    return false
  }

  public getPossiblePlays(onlyInvalid: boolean = false) {
      const availablePlay: IPlay[] = []
      const blocksToCheck = this.deterministic_mode ? range(0, 5) : [this.log.getCurrentBlock()]
      let globalIndex = 0
      for (const block of blocksToCheck) {
        const currentBlock = new Tetromino(block, 0, {
            x: 0,
            y: 0
        })
        for (const xPos of range(0, GAMEFIELD_MAX_WIDTH)) {
          // isBlocked = true
          currentBlock.position = {
              x: xPos,
              y: 0
          }
          for (const rotation of range(0, amountRotations[block])) {
            currentBlock.rotation = rotation
            const canPlace = this.canPlaceBlock(currentBlock)
            if (canPlace && !onlyInvalid || !canPlace && onlyInvalid) {
                availablePlay.push({
                    x: xPos,
                    rotation,
                    block,
                    globalIndex
                })
            }
            globalIndex++
          }
        }
      }
      return availablePlay
  }

  public placeBlockUsingPlay(play_index: number) {
    const play = this.getPossiblePlays()[play_index]
    if (play != null) {
      return this.placeBlockWithPlay(play)
    }
    return false
  }

  public placeBlockWithPlay(play: IPlay) {
    const blockType = play.block
    const y = this.getHighestY(play.x, new Tetromino(blockType, play.rotation, {
      x: play.x,
      y: 0
    }))
    const block = new Tetromino(blockType, play.rotation, {
      x: play.x,
      y
    })
    block.from = this.log.getCurrentPlayer()
    return this.placeBlock(block)
  }

  public placeBlock(block: Tetromino) {
    if (this.canPlaceBlock(block)) {
      this.blocks.push(block)
      this.blockRestrictions[this.log.getCurrentPlayer()][block.type]--
      this.log.blockPlaced(this.log.getCurrentPlayer(),
        block.type, block.rotation, block.position.x, block.position.y)
      this.determineNextPlayer()
      // console.log("JSON IS", JSON.stringify(jsToRustLog(this.log, this.players)), this.blockRestrictions)
      return true
    }
    return false
  }

  public getScores() {
    const score = new GameScore(this)
    const scores = []
    for (const p of this.players) {
      scores.push(score.getScoreForPlayer(p))
    }
    return scores
  }

  public placeCurrentBlock() {
    const play = this.log.getCurrentPlay()
    if (play != null) {
      return this.placeBlockUsingPlay(play)
    }
    return false
  }

  public randomNextPlayer() {
    const next = randomInRange(0, this.players.length - 1)
    this.rollBlockFor(this.players[next])
    this.log.createConsideration()
  }

  public setCurrentPlay(newIndex: number) {
    this.log.createConsideration(newIndex)
  }

  public rollBlockFor(player: string) {
    let next = this.players.indexOf(player) + 1
    if (next >= this.players.length) {
      next = 0
    }
    let rolledBlock: number = 0
    if (this.deterministic_mode) {
      this.log.rolledBlock(this.players[next], rolledBlock)
    } else {
      let playerHasBlock = false
      let loopedOnce = false
      while (!playerHasBlock) {
          if (next >= this.players.length) {
            next = 0
            if (loopedOnce) {
              this.gameOver = true
              return false
            }
            loopedOnce = true
          }
          rolledBlock = this.rollBlock()
          if (this.blockRestrictions[this.players[next]][rolledBlock] > 0) {
              playerHasBlock = true
          } else {
              this.log.playerSkipped(this.players[next], rolledBlock, skipReason.blockUnavailable)
              next++
          }
      }
      this.log.rolledBlock(this.players[next], rolledBlock)
    }
    return true
  }

  public determineNextPlayer() {
    if (this.rollBlockFor(this.log.getCurrentPlayer())) {
      if (this.getPossiblePlays().length <= 0) {
        this.log.playerSkipped(this.log.getCurrentPlayer(),
          this.log.getCurrentBlock(), skipReason.unplacable)
        this.rollIndex++
        if (this.rollIndex >= this.players.length) {
          this.gameOver = true
        } else {
          this.determineNextPlayer()
        }
      } else {
        this.rollIndex = 0
        this.log.createConsideration()
      }
    }
  }

  public rollBlock() {
    return randomInRange(0, tetromino.l + 1)
  }

  public restoreFromLog(log: GameLog) {
    this.log = log
    this.blocks = log.getBlocks()
    this.initBlockRestrictions()
    for (const player of this.players) {
      this.blockRestrictions[player] = this.log.getBlockCount(player)
    }
    if (this.getPossiblePlays().length === 0) {
      this.gameOver = true
    }
  }

  public getPlayerColor(player: string) {
    if (this.playerColors[player]) {
      return this.playerColors[player]
    }
    return "#000"
  }

  public chooseFittingPlay(left?: boolean, isMove?: boolean, switchBlock?: boolean) {
    const plays = this.getPossiblePlays()
    let play_index = this.log.getCurrentPlay()
    let currentBlock = plays[play_index].block
    let currentRotation = plays[play_index].rotation
    let currentX = plays[play_index].x
    const maxRot = amountRotations[currentBlock]
    const changer = left ? -1 : 1
    let loopedOnce = false
    while (true) {
      if (switchBlock) {
        currentBlock += changer
        if (currentBlock < 0) {
          if (loopedOnce) {return}
          loopedOnce = true
          currentBlock = 4
        } else if (currentBlock >= 5) {
          if (loopedOnce) {return}
          loopedOnce = true
          currentBlock = 0
        }
      } else if (isMove) {
        currentX += changer
        if (currentX >= GAMEFIELD_MAX_WIDTH) {
          if (loopedOnce) {return}
          loopedOnce = true
          currentX = 0
        } else if (currentX < 0) {
          if (loopedOnce) {return}
          loopedOnce = true
          currentX = GAMEFIELD_MAX_WIDTH - 1
        }
      } else {
        currentRotation += changer
        if (currentRotation >= maxRot) {
          if (loopedOnce) {return}
          loopedOnce = true
          currentRotation = 0
        } else if (currentRotation < 0) {
          if (loopedOnce) {return}
          loopedOnce = true
          currentRotation = maxRot - 1
        }
      }
      const foundIndex = findIndex(plays, {
        x: currentX,
        rotation: currentRotation,
        block: currentBlock
      })
      if (foundIndex !== -1) {
        this.setCurrentPlay(foundIndex)
        return
      }
    }
  }

  public moveBlock(left?: boolean) {
    this.chooseFittingPlay(left, true)
  }

  public rotateBlock(left?: boolean) {
    this.chooseFittingPlay(left, false)
  }
  public switchBlock(left?: boolean) {
    this.chooseFittingPlay(left, false, true)
  }

  public clone() {
    const newField = new GameField(this.players, cloneDeep(this.log))
    newField.playerColors = this.playerColors
    newField.controlType = this.controlType
    newField.deterministic_mode = this.deterministic_mode
    return newField
  }
}
