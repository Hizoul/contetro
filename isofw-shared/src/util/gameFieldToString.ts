import { GameField } from "../gameLogic/gameField"
import { GAMEFIELD_MAX_HEIGHT, GAMEFIELD_MAX_WIDTH } from "../gameLogic/globals"
import { isNumber } from "lodash"

const gameFieldToString = (gameField: GameField, includeHover?: boolean) => {
  let gameFieldString = "\n"
  const pointsInField: any[] = []
  for (let height = 0; height < GAMEFIELD_MAX_HEIGHT; height++) {
    pointsInField[height] = []
    for (let width = 0; width < GAMEFIELD_MAX_WIDTH; width++) {
      pointsInField[height][width] = "o"
    }
  }
  for (const block of gameField.blocks) {
    for (const position of block.getTranslatedBlock()) {
      pointsInField[Number(position.y)][Number(position.x)] = isNumber(block.from) ? block.from : gameField.players.indexOf(block.from)
    }
  }
  const hoverBlock = gameField.getHoverBlock()
  if (includeHover && hoverBlock != null) {
    for (const position of hoverBlock.getTranslatedBlock()) {
      pointsInField[position.y][position.x] = "H"
    }
  }
  for (let width = GAMEFIELD_MAX_HEIGHT - 1; width >= 0; width--) {
    let line = ``
    for (let height = 0; height < GAMEFIELD_MAX_WIDTH; height++) {
      line += pointsInField[width][height]
    }
    gameFieldString += `${line}\n`
  }
  return gameFieldString
}

export default gameFieldToString
