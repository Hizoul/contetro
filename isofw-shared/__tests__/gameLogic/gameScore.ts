import { random, randomInRange } from "isofw-shared/src/util"
random.setSeed(420)

import { GameField } from "isofw-shared/src/gameLogic/gameField"
import { GameScore } from "isofw-shared/src/gameLogic/gameScore"
import { Tetromino } from "isofw-shared/src/gameLogic/tetromino"
import gameFieldToString from "isofw-shared/src/util/gameFieldToString"

describe("The GameScore", () => {
  it("needs some functions to be called for maximum coverage", () => {
    const gameField = new GameField(["a", "b"])
    while (!gameField.gameOver) {
      gameField.placeBlockUsingPlay(randomInRange(0, gameField.getPossiblePlays().length))
    }
    const score = new GameScore(gameField)
    expect(score.getScoreForPlayer("a")).toMatchSnapshot("player a score")
    expect(score.getScoreForPlayerAtIndex(1)).toMatchSnapshot("player b score")
    expect(gameFieldToString(gameField)).toMatchSnapshot("gamefield snapshot")
  })
  it("an unknown tetromino returns null", () => {
    expect(new Tetromino(1, 99, {x: 1, y: 2}).getTranslatedBlock()).toBe(undefined)
  })
})
