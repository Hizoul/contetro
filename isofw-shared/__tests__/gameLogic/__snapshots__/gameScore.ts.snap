// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GameScore needs some functions to be called for maximum coverage: gamefield snapshot 1`] = `
"
o0111o11o1
00o10o1111
011o01111o
o11o0o1000
oo110o000o
ooo11oo00o
o0000oo111
ooo000010o
o1111ooo0o
oooo001o0o
ooo0011o0o
oooooo1o0o
oooooo1o0o
oooooo1o0o
1111oo110o
oo0000111o
ooo00001oo
o1111ooooo
oooo1ooo00
ooo111oo00
"
`;

exports[`The GameScore needs some functions to be called for maximum coverage: player a score 1`] = `-15`;

exports[`The GameScore needs some functions to be called for maximum coverage: player b score 1`] = `-14`;
